src/main/java/org/radarbase/appserver/service/RadarProjectService.java

```diff
@@ -1,15 +1,135 @@
+    }
+
+    @Transactional(readOnly = true)
+    public RadarProjectDto getProjectById(Long id) {
+        Optional<Project> project = projectRepository.findById(id);
+
+        if (project.isPresent()) {
+            return ConverterFactory.getProjectConverter().entityToDto(project.get());
+        } else {
+            throw new NotFoundException("Project not found with id" + id);
+        }
     }
 
-    public RadarProjectDto getProjectById(String id) {
-        return null;
+    @Transactional(readOnly = true)
+    public RadarProjectDto getProjectByProjectId(String projectId) {
+        Optional<Project> project = projectRepository.findByProjectId(projectId);
+
+        if (project.isPresent()) {
+            return ConverterFactory.getProjectConverter().entityToDto(project.get());
+        } else {
+            throw new NotFoundException("Project not found with projectId" + projectId);
+        }
+    }
+
+    @Transactional(readOnly = true)
+    public Project getProjectEntityByProjectId(String projectId) {
+        Optional<Project> project = projectRepository.findByProjectId(projectId);
+
+        if (project.isPresent()) {
+            return project.get();
+        } else {
+            throw new NotFoundException("Project not found with projectId" + projectId);
+        }
+    }
+
+    @Transactional
+    public RadarProjectDto addProject(RadarProjectDto projectDto) {
+
+        Optional<Project> project;
+        if (projectDto.getProjectId() != null && !projectDto.getProjectId().isEmpty()) {
+            project = projectRepository.findByProjectId(projectDto.getProjectId());
+        } else if (projectDto.getId() != null) {
+            project = projectRepository.findById(projectDto.getId());
+        } else {
+            throw new InvalidProjectDetailsException(projectDto, new IllegalArgumentException("At least one of 'id' or 'project id' must be supplied"));
+        }
+
+        Project resultProject;
+        if(project.isPresent()) {
+            resultProject = project.get().setProjectId(projectDto.getProjectId());
+        } else {
+            resultProject = ConverterFactory.getProjectConverter().dtoToEntity(projectDto);
+        }
+        Project project1 = this.projectRepository.save(resultProject);
+        return ConverterFactory.getProjectConverter().entityToDto(project1);
+    }
+
+    //TODO Test this as right now creating new project if changing projectId
+    @Transactional
+    public RadarProjectDto updateProject(RadarProjectDto projectDto) {
+
+        Optional<Project> project = projectRepository.findByProjectId(projectDto.getProjectId());;
+        if (project.isPresent()) {
+            throw new InvalidProjectDetailsException(projectDto, new IllegalArgumentException("The provided project ID already exists"));
+        } else if (projectDto.getId() != null) {
+            project = projectRepository.findById(projectDto.getId());
```

Comment:
Same confusion here. If I want to update an entity, then I expect a valid request should contain an `id`.
Query the item by id, then update. We cannot update anything which doesn't have a primary key. So it is ideal to expect it as part of the request. That will maintain the integrity of the data. 
If we want to query based on `projectId` which is unique, then query an object based on that and update. (here we know there is an id as ).

Reply:
Agreed. Sorry this service is my bad. I completely overlooked it (since it was the first and easiest to build).